pandasaurus_cxg/graph_generator/graph_generator.py

```diff
@@ -0,0 +1,173 @@
+                    }
+                elif inner_list[2] == "subcluster_of":
+                    inner_dict_uuid = {
+                        inner_list[0]: inner_list[1],
+                        "subcluster_of": {
+                            inner_list[3]: inner_list[4],
+                        },
+                    }
+                else:
+                    inner_dict_uuid = {
+                        inner_list[0]: inner_list[1],
+                    }
+
+                for key, value in inner_dict_uuid.items():
+                    if key not in temp_dict.keys():
+                        temp_dict.update({key: value})
+                    elif key == "subcluster_of":
+                        temp_dict.get("subcluster_of").update(value)
+
+            if temp_dict not in grouped_dict_uuid.values():
+                grouped_dict_uuid[str(uuid.uuid4())] = temp_dict
+
+        # generate a resource for each free-text cell_type annotation and cell_type_ontology_term annotation
+        cell_set_class = self.ns["CellSet"]
+        self.graph.add((cell_set_class, RDF.type, RDFS.Class))
+        for _uuid, inner_dict in grouped_dict_uuid.items():
+            resource = self.ns[_uuid]
+            self.graph.add((resource, RDF.type, cell_set_class))
+            for k, v in inner_dict.items():
+                if k == "subcluster_of":
+                    continue
+                self.graph.add((resource, self.ns[k], Literal(v)))
+
+        # add relationship between each resource based on their predicate in the co_annotation_report
+        for _uuid, inner_dict in grouped_dict_uuid.items():
+            resource = URIRef(self.ns + _uuid)
+            subcluster = self.ns["subcluster_of"]
+            for ik, iv in inner_dict.get("subcluster_of", {}).items():
+                predicate = URIRef(self.ns + ik)
+                for s, _, _ in self.graph.triples((None, predicate, Literal(iv))):
+                    self.graph.add((resource, subcluster, s))
+
+    def enrich_rdf_graph(self):
+        """
+        Enrich RDF graph with enriched DataFrame from AnndataEnricher
+
+        Returns:
+
+        """
+        # add cell_type nodes and consists_of relations
+        cl_namespace = Namespace("http://purl.obolibrary.org/obo/CL_")
+        for curie, label in self.cell_type_dict.items():
+            resource = cl_namespace[curie.split(":")[-1]]
+            self.graph.add((resource, RDFS.label, Literal(label)))
+            for s, _, _ in self.graph.triples((None, self.ns["cell_type"], Literal(label))):
+                self.graph.add((s, self.ns["consists_of"], resource))
+        # add subClassOf between terms in CL enrichment
+        for _, row in self.enriched_df.iterrows():
+            for s, _, _ in self.graph.triples((None, RDFS.label, Literal(row["s_label"]))):
+                for o, _, _ in self.graph.triples((None, RDFS.label, Literal(row["o_label"]))):
+                    self.graph.add((s, RDFS.subClassOf, o))
```

Comment:
I'm not entirely certain if any methods, other than simple_enrichment, contribute additional information to the graph. This is because those methods may involve CL terms from a subset and context that are not utilized in annotations.
We should talk about this.

Reply:
By definition all enrichment methods link to terms related to those used in annotation. 

Looking at this again, I think it is clear that I have underspecified.  I think the challenge is how we deal with flattening.  In the pipelines Anita has worked on we use ROBOT or RG to strip redundancy from the flattened graph.  If we're sticking with pure python we will need something similar to the the RG redundancy stripping algo here - which will require somer thought.

For this PR I'd suggest an MVP of building a graph based on co-annotation first.  We can then move folding in the enrichment graph to a second ticket/PR.